Add HMAC support to Tock and OpenTitan

alistair23 opened the pull request:
### Pull Request Overview

This PR is the initial attempt to add HMAC support to Tock and OpenTitan.

Currently the HMAC API is similar to the AES API.

Until https://github.com/lowRISC/opentitan/issues/1276 is addressed we just have to write all the data. Once the issue is fixed we can use interrupts to asynchronously write the data (similar to UART).

### Testing Strategy

It builds

### TODO or Help Wanted

I wanted to get some feedback on the API. It seems like the AES API isn't accessible from userspace. Do we want the same for HMAC or should I add a capsule that can expose HMAC to userspace?

### Documentation Updated

- [ ] Updated the relevant files in `/docs`, or no updates are required.

### Formatting

- [X] Ran `make formatall`.

bradjc commented on kernel/src/hil/hmac.rs:
```diff
@@ -0,0 +1,39 @@
+    /// Add data to the hmac IP.
+    /// All data passed in is fed to the HMAC hardware block.
+    /// Returns the number of bytes written on success
+    fn add_data(&self, data: &'static mut [u8], len: usize) -> Result<(), ReturnCode>;
```
```suggestion
    fn add_data(&self, data: &'static mut [u8], len: usize) -> Result<usize, ReturnCode>;
```

bradjc commented on kernel/src/hil/hmac.rs:
```diff
@@ -0,0 +1,39 @@
+pub trait HMAC<'a> {
+    /// Enable the HMAC hardware.
+    /// Must be called before any other methods
+    fn init(&self);
```
We generally want to avoid initialize methods in HILs #1035, #1112

labbott replied:
The symmetric encryption APIs have enable/disable functions, should those be reworked as well?

bradjc commented on kernel/src/hil/hmac.rs:
```diff
@@ -0,0 +1,39 @@
+pub trait Client<'a> {
+    // On success the `result` will contain the digest. On error it will contain
+    // the HMAC error value.
+    fn hash_done(&'a self, result: Result<&'a mut [u8], u32>);
```
Probably worth making the error types an enum, even if the errors are specific to this hardware implementation.

alistair23 replied:
The problem is that I can't find a list of the possible errors.

labbott commented on chips/lowrisc/src/hmac.rs:
```diff
@@ -0,0 +1,248 @@
+                7 => regs.key7.set(k),
+                _ => {}
+            }
+        }
```
Not something that needs to be done here but I noticed that this pattern of how to set the keys is common across aes implementations. Might be useful to eventually put this into a common place.

jmichelp replied:
If the `key` and `digest` registers were defined as an array, this pattern would naturally go away, isn't it?
```rust
(0x1C => wipe_secret: WriteOnly<u32>),
(0x20 => key: [WriteOnly<u32>; 8]),
(0x40 => digest: [ReadOnly<u32>; 8]),
(0x60 => msg_length_lower: ReadOnly<u32>),
```

alistair23 commented:
> Looks like it matches the HMAC documentation from OpenTitan. Are you planning on implementing test cases similar to the symmetric encryption?

I started working on a test but then hit an issue.

Should this be like the AES capsule (have a Tock kernel test) or should this API be adjusted and exposed via a driver to userspace and tested from libtock-rs?

I don't think userspace will need the HMAC, but I'm not sure on what the HMAC use case is with OT enough to say for sure.

niklasad1 commented on kernel/src/hil/hmac.rs:
```diff
@@ -0,0 +1,39 @@
+    fn hash_done(&'a self, result: Result<&'a mut [u8], u32>);
+}
+
+pub const HMAC_KEY_SIZE: usize = 16;
```
AFAIK it is recommended that the key size is the same block size of the hash function.
This naming seems a bit ambiguous but I guess it's directly related `lowrisc/hmac`?

Can we make this an associated const in `Hmac trait` to be more abstract and to work with different key sizes and hash algorithms?

Can we make it an associated type in the `Hmac` trait instead?

alistair23 replied:
Yep, I have updated this to add a function that sets the hash algorithm and key size.

niklasad1 commented on chips/lowrisc/src/hmac.rs:
```diff
@@ -0,0 +1,248 @@
+//! HMAC (Hash-based Message Authentication Code).
```
which hash algorithm is used?

alistair23 replied:
OT only supports SHA256, I have updated the comment

jrvanwhy commented:
> I started working on a test but then hit an issue.
> 
> Should this be like the AES capsule (have a Tock kernel test) or should this API be adjusted and exposed via a driver to userspace and tested from libtock-rs?
> 
> I don't think userspace will need the HMAC, but I'm not sure on what the HMAC use case is with OT enough to say for sure.

In that case, I think we should implement tests in kernelspace. We can add tests from userspace later if we expose this to userspace.

niklasad1 commented on kernel/src/hil/hmac.rs:
```diff
@@ -0,0 +1,39 @@
+    fn add_data(&self, data: &'static mut [u8], len: usize) -> Result<(), ReturnCode>;
+
+    /// Request the hardware block to generate a HMAC
+    /// This doesn't return anything, instaed the client needs to have
```
```suggestion
    /// This doesn't return anything, instead the client needs to have
```

niklasad1 commented on chips/lowrisc/src/hmac.rs:
```diff
@@ -0,0 +1,248 @@
+        (0x14 => status: ReadOnly<u32, STATUS::Register>),
+        (0x18 => err_code: ReadOnly<u32>),
+        (0x1C => wipe_secret: WriteOnly<u32>),
+        (0x20 => key0: WriteOnly<u32>),
```
Replace key0 - key7 to `[u32; 8]` instead?
The naming key0 - key8 doesn't help and it would simplify the code a bit i.e, indexing directly from an integer/loop.
What's the rationale behind this design?

The same applies for `digest0  - digest7`

alistair23 replied:
There wasn't really a rationale, just copying the registers from the documentation.

I'll convert it to an array.

jmichelp commented on kernel/src/hil/hmac.rs:
```diff
@@ -0,0 +1,39 @@
+    fn set_client(&'a self, client: &'a dyn Client<'a>);
+
+    /// Set the encryption key.
+    /// Returns `EINVAL` if length is not `HMAC_KEY_SIZE`
```
AFAICT this is not what the implementation above is actually doing. The size is never checked.

alistair23 replied:
Fixed

jmichelp commented on kernel/src/hil/hmac.rs:
```diff
@@ -0,0 +1,39 @@
+pub trait HMAC<'a> {
+    /// Enable the HMAC hardware.
+    /// Must be called before any other methods
+    fn init(&self);
```
I'm also working on using the ARM CryptoCell-310 HMAC accelerator and wanted to make sure the HIL was neither duplicating nor incompatible with the OT one.
In your HIL, how do we set the underlying hash algorithm? If the HIL is only compatible with 1 algorithm I suggest to then rename it HmacSha512 (according to the `HMAC_KEY_SIZE` value)

alistair23 replied:
OT only supports SHA256. I can add something to allow specifying the hash algorithm.

alistair23 commented:
I have pushed an update. It doesn't address all comments (I still have to do that), it does expose the functions to userspace apps. I have done some testing, but I'm having issues running libtock-rs apps so it's slow going.

labbott commented on capsules/src/hmac.rs:
```diff
@@ -0,0 +1,177 @@
+            // run
+            2 => {
+                self.hmac.set_key(self.key.take().unwrap());
+                self.hmac.set_key(self.data.take().unwrap());
```
should this be self.hmac.set_data?

alistair23 replied:
Yes, it should be

labbott commented on capsules/src/virtual_hmac.rs:
```diff
@@ -0,0 +1,104 @@
+        self.data.set(data);
+        self.data_len.set(length);
+
+        Ok(0)
```
The comment says the number of bytes written, is that supposed to be 0 here?

alistair23 replied:
It shouldn't be, thanks

alistair23 commented:
Ok, this is coming to shape now.

I pushed an update that addresses all comments.

I have a libtock-rs test case that can hash data.

The only issue now is that the hash doesn't match the one in the OpenTitan test case (for the same key and data). I'll have to investigate why they don't match.

jmichelp commented on chips/lowrisc/src/hmac.rs:
```diff
@@ -0,0 +1,263 @@
+}
+
+impl hil::hmac::HMAC<'a> for Hmac<'a> {
+    fn set_hash(&'a self, hash: HMACHash, key_size: usize) -> ReturnCode {
```
nit: if I read the code correctly, `key_size` is not really needed: it's the size of a block for the given hashing algorithm. So it can be inferred, right?

I don't know OT implementations details but for a software HMAC implementation the key is not bounded: if it's bigger than a block for the selected hashing algorithm, the key is first hashed with this algorithm to make it fit in 1 block.

phil-levis commented on kernel/src/hil/hmac.rs:
```diff
@@ -0,0 +1,45 @@
+
+    /// Set the encryption key.
+    /// Returns `EINVAL` if length is not `HMAC_KEY_SIZE`
+    fn set_key(&self, key: &AppSlice<Shared, u8>) -> ReturnCode;
```
I don't think `AppSlice` should be showing up in HIL traits. It implies you can't call it from within the kernel. 

As for a model for how a single struct can implement multiple related-but-distinct algorithms in a general way, I'd suggest looking at symmetric_encyrption; note that it does have different modes but does have a single trait for a particular block width.

alistair23 replied:
Ok, good point.

If not `AppSlice` then I think the only option is a raw pointer, or is there something else we could use? I was trying to avoid unsafe code, but I don't see a way around it.

alistair23 replied:
As for the algorithm, I was expecting that the caller would set this.

For example one app would do something like this (via exposed syscalls):
```rust
set_hash(MyHash);
set_client(callback);
set_key(MyKey);
add_data(MyData1);
add_data(MyData2);
run()
get_digest(&MyDigest);
```

then another app can do the same thing, for it's own has and data. The hmac capsule will enqueue everything until run and should enforce the separation of the apps.

At the moment there is no syscall to specify the hash, I will add that.

alistair23 replied:
I pushed an update not using `AppSlice`, let me know what you think

phil-levis commented on kernel/src/hil/hmac.rs:
```diff
@@ -0,0 +1,45 @@
+use crate::returncode::ReturnCode;
+
+#[derive(PartialEq)]
+pub enum HMACHash {
```
The approach we took on Titan [h1b/hil/digest.rs](https://github.com/google/tock-on-titan/blob/master/h1b/src/hil/digest.rs) was to have a digest engine, which you could configure to do SHA or SHA-based HMAC. The benefit of this is that you don't have to potentially add two levels of virtualization. With this API, you need to potentially virtualize on top of the hardware (in case someone wants to use it for something besides HMAC, and so needs to enqueue). This will be important if we want to enable basic SHA256 mode for the OT IP block.

alistair23 replied:
I'm not clear what you mean here.

phil-levis replied:
So, the issue is, what if two capsules/structures want to perform HMAC at the same time? You have to virtualize HMAC and enqueue them.

But what if the same piece of hardware, in addition to HMAC, does standard SHA hashes? I now have two different APIs, both of which need queueing. Some clients may want to use both pieces of functionality, so invoke both SHA and HMAC. This means they'll either need some kind of combined trait, or at least two separate client callbacks. But the semantics of the APIs will be such that you can call HMAC, have it enqueued, then also call SHA and have it enqueued.

Underneath, the two ways to solve this are either to have a complex and custom virtualization implementation (which handles the different kinds of outstanding operations), or to do two levels of virtualization: one virtualizes the narrow trait (SHA or HMAC) on top of a virtualizer for the underlying hardware implementation, which provides both traits.

Does that make more sense?

alistair23 replied:
Yeah, that makes more sense.

So I'm thinking that I could pretty much just rename the current HMAC implementation to digest. Add a new virtual_hmac capsule (pretty much like the current one) and that should be it. I'll give that a go

alistair23 replied:
I pushed an update that add two levels of virtulisation.

alistair23 commented on capsules/src/hmac.rs:
```diff
@@ -0,0 +1,347 @@
+        }
+    }
+
+    /// Query and control pin values and states.
```
Yeah, I need to go through and tidy up the commnets, I don't think they all match the actual functions.

niklasad1 commented on capsules/src/hmac.rs:
```diff
@@ -0,0 +1,362 @@
+                };
+
+                match app.data.as_ref() {
+                    Some(d) => match self.hmac.add_data(d.ptr(), d.len()) {
```
replace match with:

```rust
if let Err(e) = self.hmac.add_data(d.ptr(), d.len()) {
  return e;
}
```

niklasad1 commented on capsules/src/hmac.rs:
```diff
@@ -0,0 +1,362 @@
+        self.appid.map(|id| {
+            self.apps
+                .enter(*id, |app, _| {
+                    match app.dest.as_ref() {
```
replace with `if let Some(d) = self.hmac.add_data(d.ptr(), d.len())`

niklasad1 commented on boards/opentitan/src/main.rs:
```diff
@@ -227,6 +251,7 @@ pub unsafe fn reset_handler() {
         led: led,
         console: console,
         alarm: alarm,
+        hmac: hmac,
```
```suggestion
        hmac,
```

niklasad1 commented on kernel/src/hil/hmac.rs:
```diff
@@ -0,0 +1,39 @@
+    fn set_client(&'a self, client: &'a dyn Client<'a>);
+
+    /// Set the hash key
+    fn set_key(&self, key: *const u8, len: usize) -> ReturnCode;
```
hmm, why is key is a raw pointer here instead of slice?

alistair23 replied:
It should be converted to a slice.

bradjc commented on chips/lowrisc/src/hmac.rs:
```diff
@@ -0,0 +1,257 @@
+    registers: StaticRef<HmacRegisters>,
+
+    client: OptionalCell<&'a dyn hil::hmac::Client<'a>>,
+    data: Cell<Option<*const u8>>,
```
You can use an `OptionalCell` if you want.

lschuermann commented on kernel/src/hil/hmac.rs:
```diff
@@ -0,0 +1,39 @@
+    fn hash_done(&'a self, result: Result<(), u32>);
+}
+
+pub trait HMAC<'a> {
```
I think we could probably make this trait generic over the underlying hash function and thereby avoid having to to set a `SHA256` mode and can return a fixed-size buffer with length determined at compile time with `get_digest()`. I could build something like that as a PR to your fork if you'd like.

alistair23 replied:
How could we pass that fixed sized buffer to userspace though?

lschuermann replied:
Passing to userspace is an issue I'm not particularly concerned with in the HIL. Please have a look at #1727 to see what I mean by abstracting over a generic type. Still waiting for feedback on that though.

lschuermann commented on kernel/src/hil/hmac.rs:
```diff
@@ -0,0 +1,39 @@
+    /// Add data to the hmac IP.
+    /// All data passed in is fed to the HMAC hardware block.
+    /// Returns the number of bytes written on success
+    fn add_data(&self, data: *const u8, length: usize) -> Result<usize, ReturnCode>;
```
This may be an asynchronous operation on many hardware platforms and therefore should return a callback to the client. For efficiency reasons it may be desirable to not keep the scheduler busy with callbacks, so you could use the return value to indicate whether a client must wait for a callback.

alistair23 replied:
It is asyncronous here (not in the actual OT implementation due to a bug though). We call-back internally in the driver, kind of similar to the UART implementation.

lschuermann commented on kernel/src/hil/hmac.rs:
```diff
@@ -0,0 +1,39 @@
+    fn run(&'a self);
+
+    /// Stores the returned digest in the memory location specified
+    fn get_digest(&self, digest: *mut u8, len: usize) -> ReturnCode;
```
Why can't we just provide the hash to the client in the `hash_done()` callback?

alistair23 replied:
I thought about that, but I was worried the app would not just want the data in the callback. This was it's more flexible.

lschuermann commented on kernel/src/hil/hmac.rs:
```diff
@@ -0,0 +1,39 @@
+    /// Stores the returned digest in the memory location specified
+    fn get_digest(&self, digest: *mut u8, len: usize) -> ReturnCode;
+
+    /// Disable the HMAC hardware and clear the keys and any other sensitive
```
I wouldn't make the comment indicate that the hardware will be disabled, for instance that'd be false for software implementations. Hardware implementations may be designed in a way that they automatically turn off the HW on a clear and turn it on again on data input, however that should either be done transparent to the client, or as extra methods in the trait which may fail - depending on actual support.

alistair23 replied:
Good point, I'll remove this

alistair23 commented:
I have converted the `set_key()` function to use a slice, that is easy.

For the `add_data()` function it's harder due to Rust lifetimes. I ended up copying the data (which I don't really like. Let me know if anyone knows of a better way of async writing the data while not using a static chunk of memory.

I have also converted `get_digest()`, it is also straightforward.

bradjc commented on capsules/src/hmac.rs:
````diff
@@ -0,0 +1,358 @@
+//! digest::Digest::set_client(virtual_hmac_user, hmac);
+//! ```
+
+/// Syscall driver number.
````
```suggestion
```

bradjc commented on capsules/src/hmac.rs:
```diff
@@ -0,0 +1,358 @@
+
+/// Syscall driver number.
+use crate::driver;
+pub const DRIVER_NUM: usize = driver::NUM::Hmac as usize;
```
```suggestion
/// Syscall driver number.
pub const DRIVER_NUM: usize = driver::NUM::Hmac as usize;
```

bradjc commented on capsules/src/virtual_hmac.rs:
```diff
@@ -0,0 +1,105 @@
+use kernel::hil::digest;
+use kernel::ReturnCode;
+
+pub struct VirtualMuxHmac<'a, A: kernel::hil::digest::HMACSha256 + digest::Digest<'a>> {
```
Why not call this `HmacDevice` to mirror the other virtual_ capsules?

bradjc replied:
Oh I see because of virtual_alarm example. Hmm. I don't have an intuitive idea what a `VirtualMuxX` is, but maybe it makes sense to others.

alistair23 replied:
Yeah, I based it on the virtual_alarm. Let me know if you want it changed.

bradjc commented on capsules/src/virtual_digest.rs:
```diff
@@ -0,0 +1,106 @@
+use kernel::hil::digest;
+use kernel::ReturnCode;
+
+pub struct VirtualMuxDigest<'a, A: kernel::hil::digest::HMACSha256 + digest::Digest<'a>> {
```
Why not `DigestDevice`?

alistair23 replied:
Same as above

bradjc commented on kernel/src/hil/digest.rs:
```diff
@@ -0,0 +1,38 @@
+    fn hash_done(&'a self, result: Result<(), u32>);
+}
+
+pub trait Digest<'a> {
```
What does this do? As in what would a user expect to happen if they used a `Digest` interface?

phil-levis replied:
It computes a digest (a cryptographic hash) over data.

alistair23 replied:
A user shouldn't use the digest interface, the idea is that a `Hmac` virtual device is implemented on top of a `Digest`. This would also allow a `SHA` virtual device on top of the `Digest` and the same hardware can be supported for both (if the HW supports both).

phil-levis commented on kernel/src/hil/digest.rs:
```diff
@@ -0,0 +1,38 @@
+    fn set_client(&'a self, client: &'a dyn Client<'a>);
+
+    /// Set the hash key
+    fn set_key(&self, key: &[u8], len: usize) -> ReturnCode;
```
Technically a digest (cryptographic hash) does not have a key: HMAC using a digest does, however. So I'd suggest pulling `set_key` out of this trait and instead having `set_mode_hmacsha256` take the key. This would also allow you to statically check the key length by specifying it as a reference to an array.

alistair23 replied:
Good idea, fixed in the next version.

alistair23 commented:
>     1. I'm not keen on merging the userspace capsule, as it puts us further in debt towards #1216. HMAC seems like something that should be sharable among multiple processes.

Agreed that it should be shareable. Isn't that what the virtual_hmac does? I haven't tested this between multiple apps, but in theory shouldn't it be shareable?

Also, I could always drop the userspace capsule and that could be a followup PR.

>     2. Is it helpful to have both virtual_digest and virtual_hmac? I guess I'm not clear on what the Digest interface does.

That was @phil-levis suggestion. I have copied his explanation below as I think he does a good job of explaining it:

> So, the issue is, what if two capsules/structures want to perform HMAC at the same time? You have to virtualize HMAC and enqueue them.
> 
> But what if the same piece of hardware, in addition to HMAC, does standard SHA hashes? I now have two different APIs, both of which need queueing. Some clients may want to use both pieces of functionality, so invoke both SHA and HMAC. This means they'll either need some kind of combined trait, or at least two separate client callbacks. But the semantics of the APIs will be such that you can call HMAC, have it enqueued, then also call SHA and have it enqueued.
> 
> Underneath, the two ways to solve this are either to have a complex and custom virtualization implementation (which handles the different kinds of outstanding operations), or to do two levels of virtualization: one virtualizes the narrow trait (SHA or HMAC) on top of a virtualizer for the underlying hardware implementation, which provides both traits.

>     3\. The HIL file needs more comments explaining the intended structure.

I'll add some more comments then.

> Also, have you looked at #1727? I'm not sure the best path forward. Although I am happy to merge this version of the HIL and then use #1727 as a chance to update.

Yeah, they just submitted a PR. It looks similar, but it doesn't have an implementation or all the virtulisation required.

alistair23 commented:
I have pushed an update addressing comments.

It's using the two level virtulisation.
I have added some more comments to the `Digest` HIL to make it clearer.
I have removed the `set_key()` call from the `Digest` HIL and replaced it with setting the key in the `HMACSha256`.

As for OT DIFs, I don't see why this wouldn't be compatible. There are only so many ways to generate HMACs :) I have no insight into what is going on with DIFs though so I can't really comment.

lschuermann commented:
> I have removed the set_key() call from the Digest HIL and replaced it with setting the key in the HMACSha256.

I _think_ I know where you're going with this and the two-level virtualisation, however I believe it'd be nicer to have a consolidated HMAC trait which has all methods required. You might want to have HW which can _only_ do HMAC, and I believe this isn't possible with this setup.

I that way I am probably going against @phil-levis argument of the two-client usage scenario, but sharing resources isn't easy even when you're only exposing one trait like SHA. Even then you must consider multiple clients wanting to access that particular resource. The two-level virtualisation is particularly effective when you've got one client for hashing, one for HMAC. With multiple clients for both, this appears to be double the effort for me and I'd prefer having virtualisation located "lower down", behind those traits, as we'd need it there anyways.

I had some really old code lying around for a hash abstraction (not having HMAC) in mind, which uses generics instead of methods to set a particular hash function. With you're generalization of the HMAC trait into Digest we've run into a "race condition" of two things with overlapping use cases, which is not what I intended. Still, I think that the abstraction over generic types in #1727 could help here too, especially if this were an `HMAC` instead of `Digest` trait.

phil-levis commented on kernel/src/hil/digest.rs:
```diff
@@ -0,0 +1,49 @@
+    ///
+    /// The key used for the HMAC is passed to this function as well
+    /// as the length of the key data.
+    fn set_mode_hmacsha256(&self, key: &[u8], len: usize);
```
Should this be a fixed, known keysize (i.e., &[u8; 16])?

jmichelp replied:
HMAC doesn't limit the size of the key. If it's bigger than a block for the selected underlying digest, the provided key is hashed before being used

jmichelp replied:
But providing the length is redundant. The slice already embeds it.

alistair23 replied:
I have left this as not a known size and removed the length argument.

alistair23 commented:
I'm happy to have patches on top of this for improvements if you want @lschuermann

alistair23 commented:
What are peoples thoughts on this?

If we want improvements on top based on https://github.com/tock/tock/pull/1727 that can happen after this PR is merged?

lschuermann commented:
> I'm happy to have patches on top of this for improvements if you want @lschuermann

@alistair23 I'm honestly pretty happy with how this is coming together. As our HIL's are quite similar indeed, I think this also validates my ideas and will make merging #1727 even easier. If you don't mind I'd give this an in-depth review over the next 2 days and send patches if I see something. Amazing work, glad crypto becomes a focus area of Tock! :+1: 

If I were to criticize on thing, it'd be that this PR is huge. It's no ones fault, just evolved that way. However that makes nitpicking on the trait a bit harder than I'd like. :sweat_smile: 
I think the actual HW crypto implementation is probably fine, and this is the primary focus of this PR. I'd love to revisit the APIs in #1727 and look for improvements to the HILs together. That isolates that part of the problem and should make it a lot easier to deal with. :smiley:

alistair23 commented:
Great! Thanks.

This PR is split up into small patches, unfortunately the GitHub UI is pretty useless at looking at actually commits.

lschuermann commented on kernel/src/hil/digest.rs:
```diff
@@ -0,0 +1,48 @@
+    /// The data can be fed to the underlying hardware block either syncronously
+    /// or asyncronously. This is up to the hardware implementation and is
+    /// transperent to the user.
+    fn add_data(&self, data: &[u8]) -> Result<usize, ReturnCode>;
```
Up to my knowledge it's not possible to transparently abstract over (a)synchronicity without busy-waiting. Either the component does it directly, or you'll have to wait - if this were transparent that would have to happen in the same call stack.

Therefore, a call to `add_data` in my opinion should incur a callback. If for efficiency reasons you'd like to avoid going down to the scheduler again, you could also return a boolean indicating whether a callback will be fired (asychronous operation) and hence the client would have to wait on that.

alistair23 replied:
Agreed! I'll add a callback.

phil-levis replied:
You can't always issue a callback unless you use deferred procedure calls. This adds a lot of complexity. The general programming idiom we use in Tock is to return a `(ReturnCode, Option<&[u8]>)` so you don't lose the reference to the buffer. E.g., 

https://github.com/tock/tock/blob/b134ba078c3d80a511dbd42aa05d0ff908bdd594/kernel/src/hil/uart.rs#L109

https://github.com/tock/tock/blob/b134ba078c3d80a511dbd42aa05d0ff908bdd594/kernel/src/hil/radio.rs#L103

Not all HILs have adopted this approach yet.

lschuermann replied:
@phil-levis I did not know about this approach. I would have used a deferred call here, but this seems nice as it does not unnecessarily go down to scheduler. Nonetheless, there needs to be the possibility of a callback if it were an asynchronous operation (which I guess in this instance should be the case quite often), this captures both these two options.

alistair23 replied:
Done!

lschuermann commented on kernel/src/hil/digest.rs:
```diff
@@ -0,0 +1,48 @@
+
+    /// Stores the returned digest in the memory location specified. This
+    /// is the only way to get the output of the digest.
+    fn get_digest(&self, digest: &mut [u8]) -> ReturnCode;
```
This is unnecessary. You should instead pass a slice in the `hash_done` callback. Having this method only introduces more code paths that could be erroneous: a client can call `get_digest()` even when the digest is not ready. Also a client would be able to call `get_digest()` `n`-times, requiring the implementer to handle these scenarios. It's much easier if the generated digest is given once in `hash_done` - which given a correct implementation is guaranteed to be only called as a result of a successful operation, and only once.

alistair23 replied:
The only problem then is that we need some call that specifies where to save the digest. I guess the best option is to add an argument to `run()` and then remove this function. I can do that.

lschuermann replied:
That is an option, though not my favorite. I get why you can't hand out a simple reference to an `u8` slice in the callback - internally you're working with `u32`s. But bear in mind that passing the buffer in the callback does not require any persistent allocation. It is enough to pass in a reference to the buffer on the call stack. As that isn't too big, I think it should be preferred here. The client would need to copy the contents in the callback function anyways.

A viable alternative could be to pass a `&mut dyn Iterator<type = u8>` to the callback. That way you never have to allocate the full buffer and can build the `u8` iterator from your `u32`s on the fly. However that introduces some level of complexity - it may be easier to have a stack allocated temporary buffer as proposed above.

lschuermann commented on kernel/src/hil/digest.rs:
```diff
@@ -0,0 +1,48 @@
+    /// set a `hash_done` handler to determine when this is complete.
+    /// If there is data from the `add_data()` command asyncrously waiting to
+    /// be written it will be written before the operation starts.
+    fn run(&'a self);
```
In my proposed changes to the HIL I've changed the return type to `Result<(), ReturnCode>`, giving this the option to fail fast. It isn't stricly necessary, but can indicate the client that something's off without it having to wait on the callback.

alistair23 replied:
I'm expecting that the caller will be more interested in getting the result in the callback and that this wouldn't add a lot. I'm open to adding it though

lschuermann replied:
It doesn't add a lot and is fine this way. I started to think about what reasons could be making it fail fast. Those probably don't add up to the value of having this return `Result`.

alistair23 replied:
I added a result here

lschuermann commented on kernel/src/hil/digest.rs:
```diff
@@ -0,0 +1,48 @@
+/// Implement this trait and use `set_client()` in order to receive callbacks.
+pub trait Client<'a> {
+    /// On error `result` will contain device specific error codes
+    fn hash_done(&'a self, result: Result<(), u32>);
```
A HIL should abstract the HW. Having device specific error codes doesn't gain anything. It'd be better to have a mapping of error scenarios to an enum such as `ReturnCode`, but I don't yet know of any convention here. I always choose a variant which _feels right_.

In line with my other comment, this should provide the calculated hash.

alistair23 replied:
I would prefer to have specific errors, the problem is that I don't have a list of possible OpenTitan errors, and even if I did I don't have a list of other module errors. So it seems hard to generalise this to a small list of errors. My thinking is that during operation we aren't really expecting errors (what can fail?) and that this is just for debugging.

lschuermann replied:
Still, this is a HIL. This way, it would be very incompatible with any other HW being interfaced with. If you can't find a list, and don't know what could go wrong, I'd suggest always returning `ReturnCode::FAIL` if there has been an error - the client won't know how to interpret the `u32` either.

lschuermann commented on kernel/src/hil/digest.rs:
```diff
@@ -0,0 +1,48 @@
+use crate::returncode::ReturnCode;
+
+/// Implement this trait and use `set_client()` in order to receive callbacks.
+pub trait Client<'a> {
```
This should have additional callbacks for
- `clear_data`
- `add_data`

as these could be asynchronous depending on the HW. Maybe, a callback for setting the mode (and key) would be a good idea as well. If these are not asynchronous with OpenTitan, consider using a deferred call.

alistair23 replied:
Do we need a callback for clear data? What action are we expecting to do on the callback?

lschuermann replied:
Not necessarily an action, just an indication that the possibly long-running and asynchronous operation of clearing data or resetting the device has finished, and the client may proceed to calculate digests again.

alistair23 replied:
I haven't added a callback for `clear_data()` I'm not convinced we need one. Let me know what you think of the latest patches.

lschuermann replied:
I'm convinced one will be required sooner or later. Fine with me though, as soon as some HW comes around requiring it it must be added.

lschuermann commented on kernel/src/hil/digest.rs:
```diff
@@ -0,0 +1,48 @@
+    fn get_digest(&self, digest: &mut [u8]) -> ReturnCode;
+
+    /// Clear the keys and any other sensitive data.
+    /// This won't clear the buffers provided to this API, that is up to the
```
What buffers? As far as I understand, the user only provides references to buffers which are valid for the individual calls.

alistair23 replied:
The user/app allocates buffers with the key, data and digest. These are passed as references to the digest HIL. I'm just making it clear that these won't be cleared and that's up to the user/app to do.

lschuermann replied:
Fine. However that is already guaranteed by the type signature (lifetimes) of the HIL. In it's current state the `clear_data` function could never (without resorting to unsafe code) "clear" any buffer that's just been handed in as a reference to a function. For this, the HIL would need to "own" the mutable reference to the buffer (for instance in `get_digest`), but that doesn't happen, as the automatically derived lifetimes will "hand that mutable buffer reference back" after the function execution (please correct me if I'm wrong here).

alistair23 replied:
In the updated version we expose `mut` versions of the buffers, so we could clear them. I'm just making sure it's explicit that we won't clear them.

alistair23 commented:
Updated! Based on comments from @phil-levis and @lschuermann

bradjc commented on kernel/src/hil/digest.rs:
```diff
@@ -0,0 +1,62 @@
+    /// There is no guarentee the data has been written until the `add_data_done()`
+    /// callback is fired.
+    /// On error the return value will contain a return code and the original data digest
+    fn add_data(&self, data: &'static mut [u8]) -> Result<usize, (ReturnCode, &'static mut [u8])>;
```
I think this needs to take a length as well (or use a LeaseableBuffer https://github.com/tock/tock/blob/master/kernel/src/common/leasable_buffer.rs). The issue is if a user wants to add only a portion of a buffer to a digest, it either has to have a buffer exactly that length, or provide only a slice to Digest. If it only provides a slice, then it will only ever get back that slice, and looses access to the entire buffer.

jmichelp replied:
I don't see why the `data` slice should be mutable here.
And if it isn't mutable, the caller can pass a slice without worrying about passing an extra length nor loosing the slice.

bradjc replied:
But then the caller wouldn't be able to hold on to the mutable reference as well. `data` is `&mut` so that the HMAC implementor can pass back a mutable reference when it is done with the buffer.

alistair23 replied:
I have updated this to use `LeasableBuffer`

bradjc commented on kernel/src/hil/digest.rs:
```diff
@@ -0,0 +1,62 @@
+    /// Add data to the digest block. This is the data that will be used
+    /// for the hash function.
+    /// Returns the number of bytes parsed on success
+    /// There is no guarentee the data has been written until the `add_data_done()`
```
```suggestion
    /// There is no guarantee the data has been written until the `add_data_done()`
```

bradjc commented on capsules/src/virtual_hmac.rs:
```diff
@@ -0,0 +1,104 @@
+    /// Returns the number of bytes written on success
+    fn add_data(&self, data: &'static mut [u8]) -> Result<usize, (ReturnCode, &'static mut [u8])> {
+        self.mux.hmac.add_data(data)
+    }
```
I don't think this actually virtualizes, does it? Wouldn't two `VirtualMuxHmac`s using this interface at the same time interleave their data?

alistair23 replied:
Yep :(

alistair23 replied:
I had a go at actually muxing this. I can't think of a good way to mux the key selection. Any ideas?

alistair23 replied:
I added a basic only one user at a time check. I couldn't think of a nice way to support multiple users at the same time as we would need to buffer data which I was trying to avoid.

bradjc commented on kernel/src/hil/digest.rs:
```diff
@@ -0,0 +1,62 @@
+    /// On error the return value will contain a return code and the original digest
+    /// If there is data from the `add_data()` command asyncrously waiting to
+    /// be written it will be written before the operation starts.
+    fn run(&'a self, digest: &'static mut T) -> Result<(), (ReturnCode, &'static mut T)>;
```
I'm still curious what should happen if a user calls `run` without ever calling `set_mode_hmacsha256()`.

alistair23 replied:
Right now it would just work I think as it's the only option.

bradjc replied:
What happens if there are two options?

I think conceptually I like the `Digest<DigestType>` approach from #1727 more, primarily because this question doesn't exist, and there is no confusion about what it means to directly use a `Digest` interface. However, practically I think the HmacAlgorithm::set_mode() approach in this PR is more flexible and usable and leads to easier to read code, particularly with a single hardware peripheral that can support numerous modes.

But, given this is a HIL, it should describe what an implementer is supposed to do if a user uses it incorrectly (i.e. doesn't call `set_mode_X()`). Should it just pick one? Does it return an error (which error)? Does it panic?

alistair23 replied:
I added some documentation describing what should be done.

gendx commented on capsules/src/hmac.rs:
```diff
@@ -0,0 +1,503 @@
+                                        // Copy the entire buffers worth of data
+                                        buf.copy_from_slice(
+                                            d.as_ref()
+                                                .get(copied_data..(copied_data + static_buffer))
+                                                .unwrap(),
```
Same a above regarding `get` + `unwrap`. I don't even think you need the `as_ref` part.

gendx commented on capsules/src/hmac.rs:
```diff
@@ -0,0 +1,503 @@
+                                    self.data_buffer
+                                        .take()
+                                        .unwrap()
+                                        .get_mut(0..(data_len - copied_data))
+                                        .unwrap(),
```
Same regarding get + unwrap.

gendx commented on capsules/src/hmac.rs:
```diff
@@ -0,0 +1,503 @@
+
+                    if let Err(_e) = self.hmac.run(self.dest_buffer.take().unwrap()) {
+                        app.callback.map(|cb| {
+                            // cb.schedule(e.0 as usize, e.1, 0);
```
What is this commented line for?

alistair23 replied:
I have fixed this, I was meant to call the commented function.

alistair23 commented on chips/lowrisc/src/hmac.rs:
```diff
@@ -0,0 +1,250 @@
+
+                let data_idx = i * 4;
+
+                let mut d = (slice[data_idx + 0] as u32) << 0;
+                d |= (slice[data_idx + 1] as u32) << 8;
+                d |= (slice[data_idx + 2] as u32) << 16;
+                d |= (slice[data_idx + 3] as u32) << 24;
```
`from_be_bytes()` won't work on a slice though. The example in the Rust docs uses `split_at()` to fix this, but that will change `slice` which we can't do as we need to pass it back to the callback.

alistair23 commented on capsules/src/hmac.rs:
```diff
@@ -0,0 +1,503 @@
+                                    // Copy the data into the static buffer
+                                    if data_len < (copied_data + static_buffer) {
+                                        // If the static buffer is larger then the data, just copy the rest of the slice
+                                        for (i, d) in d
```
Unfortunately it can't be. `copy_from_slice()` requires the lengths to be the same, in this case the two aren't the same length.

gendx replied:
I meant something along the lines of `foo[a..b].copy_from_slice(&bar[c..d])`. Then the lengths would match, right?

alistair23 commented on kernel/src/hil/digest.rs:
```diff
@@ -0,0 +1,62 @@
+pub trait Client<'a> {
+    /// This callback is called when the data has been added to the digest
+    /// engine.
+    /// On error `result` will contain a return code
```
I'm not sure what you mean.

gendx replied:
Sorry, it wasn't clear to me that these were callbacks, with `result` and `data` being *inputs* to this function. The "will contain" phrasing (in the future tense) suggests that `result` and `data` are populated by this function, and contain foo/bar after the function is called.

lschuermann replied:
I think the description is fine and in the context of callbacks using future tense, especially combined with the type signature is what I would have expected.

alistair23 commented on kernel/src/hil/digest.rs:
```diff
@@ -0,0 +1,62 @@
+impl DigestType for [u8; 32] {}
+
+/// Computes a digest (cryptographic hash) over data
+pub trait Digest<'a, T: DigestType> {
```
Based on the compiler errors I'm seeing I still need the `T` trait even with your example code.

I have added some comments to try and clear things up, let me know what you think.

gendx replied:
My comment was about renaming these traits and types to more meaningful names - not changing the actual programming logic.

lschuermann replied:
I think `FooType` is a good trait name scheme, especially when used to _configure_ another component avoiding confusion due to similar names. Bad example incoming: When implementing a `Driver`, you're not naming the trait `DriverTrait`, but rather after the thing it's going to be when implemented. In this case a `DigestType`.

What I do find awkward (I'm biased) is calling an `[u8; 32]` a _DigestType_, as that's rather just the hash function's output type. This doesn't distinguish between SHA256 and SHA3-256, which are very different hash functions (my definition of a `HashType`, could also be called `HashFunction`) producing the same output type. But changing it to that introduces so much more complexity, which we should preferably look at those suggestions in #1727.

gendx replied:
To me, saying that the output of a hash is a "digest" makes sense. A service providing some hashing facility is not much of a digest itself, maybe more of a "hasher". Maybe the following renaming would be clearer:

```
Digest => Hasher
DigestType => Digest (the trait)
T => DigestType (the actual type)
```

```rust
pub trait Hasher<'a> {
    type DigestType: Digest;

    fn run(&'a self, digest: &'static mut DigestType) -> Result<...>;
}
```

gendx commented on kernel/src/hil/digest.rs:
```diff
@@ -0,0 +1,68 @@
+    fn add_data_done(&'a self, result: Result<(), ReturnCode>, data: &'static mut [u8]);
+
+    /// This callback is called when a digest is computed.
+    /// On error `result` will contain a return code
```
This sentence doesn't add information IMO. The type of `result` is `Result<(), ReturnCode>` so it's clear that it contains a `ReturnCode` in case of `Err`.

gendx commented on capsules/src/hmac.rs:
```diff
@@ -0,0 +1,492 @@
+                                // Determine how much data we have already copied
+                                let copied_data = self.data_copied.get();
+
+                                data_len = d.as_ref().len();
```
Same as above regarding `d.as_ref()`.

gendx commented on capsules/src/hmac.rs:
```diff
@@ -0,0 +1,492 @@
+                                // If we have more data then the static buffer
+                                if data_len > copied_data {
+                                    // Copy the data into the static buffer
+                                    if data_len < (copied_data + static_buffer) {
```
There could be an integer overflow in `copied_data + static_buffer`. How are we making sure that's not the case?

The following would be clearer IMO:

```rust
let remaining_data = &d.as_ref()[copied_data..];
let remaining_len = data_len - copied_data;

if remaining_len < static_buffer {
    buf[..remaining_len].copy_from_slice(remaining_data);
} else {
    buf.copy_from_slice(&remaining_data[..static_buffer]);
}
```

gendx commented on capsules/src/hmac.rs:
```diff
@@ -0,0 +1,492 @@
+                        match app.data.as_ref() {
+                            Some(d) => {
+                                // Determine the size of the static buffer we have
+                                static_buffer = buf.len();
```
The naming of this variable is unfortunate. It represents a length (`usize`), not a buffer (array or slice).

gendx commented on capsules/src/hmac.rs:
```diff
@@ -0,0 +1,492 @@
+
+                            // Add the data from the static buffer to the HMAC
+                            if data_len < (copied_data + static_buffer) {
+                                lease_buf.slice(0..(data_len - copied_data))
```
`0..a` can be re-written as `..a`.

bradjc commented on capsules/src/virtual_digest.rs:
```diff
@@ -0,0 +1,136 @@
+
+pub struct VirtualMuxDigest<
+    'a,
+    A: kernel::hil::digest::HMACSha256 + digest::Digest<'a, T>,
```
```suggestion
    A: digest::Digest<'a, T>,
```

If this virtualizes only on Digest, it shouldn't assume the Digest hardware supports a particular algorithm, correct?

bradjc commented on capsules/src/virtual_digest.rs:
```diff
@@ -0,0 +1,136 @@
+pub struct MuxDigest<'a, A: kernel::hil::digest::HMACSha256 + digest::Digest<'a, T>, T: DigestType>
+{
+    digest: &'a A,
+    running: Cell<bool>,
```
```suggestion
    running: OptionalCell<usize>,
```

I think keeping track of which `VirtualMuxDigest` is running is the missing piece to enabling virtualization. Then only that client can use it until the digest engine is no longer running.

alistair23 replied:
I couldn't figure out a good way to keep track of which one is running. Any ideas?

Currently we just rely on the one that starts to say that it is running.

bradjc replied:
My first thought would be to assign an "id" when the virtual client is created, by having the mux increment a counter every time a new client is created. Then the mux can use that id to mark who currently owns the digest HW.

alistair23 commented:
I have pushed an update addressing most things. Let me know what everyone thinks

gendx commented on boards/components/src/hmac.rs:
````diff
@@ -0,0 +1,148 @@
+//! -----
+//! ```rust
+//! let led = components::led::HmacsComponent::new().finalize(components::led_component_helper!(
+//!     (&nrf52840::gpio::PORT[LED_RED_PIN], kernel::hil::gpio::ActivationMode::ActiveLow),
+//!     (&nrf52840::gpio::PORT[LED_GREEN_PIN], kernel::hil::gpio::ActivationMode::ActiveLow),
+//!     (&nrf52840::gpio::PORT[LED_BLUE_PIN], kernel::hil::gpio::ActivationMode::ActiveLow)
````
How are LEDs related to this component?

phil-levis commented on kernel/src/hil/digest.rs:
```diff
@@ -0,0 +1,72 @@
+    /// Call before `Digest::run()` to perform HMACSha256
+    ///
+    /// The key used for the HMAC is passed to this function.
+    fn set_mode_hmacsha256(&self, key: &[u8]) -> Result<(), ReturnCode>;
```
This should take an array of the keysize. One issue that came up on the OpenTitan call is a concern about hardware that allows you to specify arbitrary length keys that are then hashed. This should be a separate API -- it is not an HMACSha256, it's a KDF followed by HMAC.

alistair23 replied:
Fixed

lschuermann replied:
I believe this to not be entirely true. HMAC ([RFC 2104, 2. Definition of HMAC](https://tools.ietf.org/html/rfc2104#section-2)) explicitly states that "[the] authentication key K can be of any length up to B, the block length of the hash function". While the padding is well defined (all zeros), this shouldn't be propagated to the client: IMHO to be closely following the standard the trait should accept at least any key length up to `K`, which a fixed length array would make impossible (without an additional length attribute).

Going further, under [3. Keys](https://tools.ietf.org/html/rfc2104#section-3), the reference explicitly states that "The key for HMAC can be of any length (keys longer than B bytes are first hashed using H)", which leads me to believe that accepting a key `> K` is an integral part of a standards compliant HMAC implementation.

lschuermann replied:
A good middle ground would be a trait which is standards compliant by allowing any key size, always padding if necessary, but not necessarily hashing keys longer than `K`. If a long key is passed and hashing (the KDF part) is not supported (possibly not conforming to the original standard RFC 2104), return `ReturnCode::ESIZE`, indicating that hashing is required first. This exact required behavior would need to be documented in the trait.

alistair23 replied:
What are the thought here? I kind of think a fixed key length makes the most sense as Phil says it means every implementation will do the same thing. To pad with zeros or hash the key (KDF) we should add some other call that makes it explicit. Otherwise every Digest implementation needs to make sure it handles the keys correctly, which they might not. If we split it out we can just have one common place that does the key changes in common code..

phil-levis replied:
I think the best approach is to have a fixed-length key, but then you can always layer something more flexible on top. Having a fixed size key allows you to avoid runtime error checking. If the call takes a slice, then you want to return an error if, say, the key is of length 0.  Moving to slice for a key will also require that setting the key has a callback, because computing the hash does. The call stops being synchronous.

So I think the right middle ground is to keep the HMACSha256 as is, but we can have later another trait, HMACSha256Hashed, whose set_key method takes a slice and which issues a callback.

Incidentally, should it be HmacSha256?

lschuermann replied:
I do see the advantages a fixed size key has regarding implementation complexity.

I'm also certain having another trait (`HMACSha256Hashed`) will lead to that not being implemented / used at least some of the time. We'd then have a widely accessible kernel-internal (and not crate internal etc.) interface (`HMACSha256`) which sort of implements HMAC, is _called_ HMAC, but really doesn't adhere to standards.

@phil-levis If I understand you correctly, when having a piece of hardware not supporting KDF it would implement `HMACSha256`. HW supporting that would implement `HMACSha256Hashed`. When we now start writing clients using `HMACSha256`, those will be inherently incompatible with `HMACSha256Hashed`, because a HW can't implement _both_ traits as we'd then violate the idea of the single data path in the crypto traits (which still needs discussion). The only solution I'd be comfortable with is immediately writing a wrapper struct for `HMACSha256`s, exposing them as `HMACSha256Hashed` so clients can exclusively use that trait (which would allow us to add HW later that does the KDF implicitly). That relies on a hasher though.

It's a difficult discussion - I am not the one to decide. Deliberately exposing interfaces not adhering to the standards could hurt us in the long run, so I'd like to be careful about that.

lschuermann commented on kernel/src/hil/digest.rs:
```diff
@@ -0,0 +1,72 @@
+    /// Call before `Digest::run()` to perform HMACSha256
+    ///
+    /// The key used for the HMAC is passed to this function.
+    fn set_mode_hmacsha256(&self, key: &[u8; 32]) -> Result<(), ReturnCode>;
```
This should take variable length keys as described [here](https://github.com/tock/tock/pull/1702#discussion_r409694688). The GitHub UI won't allow me to re-open the outdated review thread.

alistair23 replied:
I have un-resolved the issue, let's discuss this above.

lschuermann commented on kernel/src/hil/digest.rs:
```diff
@@ -0,0 +1,72 @@
+    /// engine.
+    /// On error or success `data` will contain a reference to the original
+    /// data supplied to `add_data()`.
+    fn add_data_done(&'a self, result: Result<(), ReturnCode>, data: &'static mut [u8]);
```
If we're already forcing the client to use a `LeasableBuffer` in the `add_data()` method, wouldn't it be reasonable to return the same type here as well?

alistair23 replied:
I don't think so as the client doesn't really want to use a `LeasableBuffer`. The  `LeasableBuffer` is just equivalent to passing in the length.

lschuermann commented on chips/lowrisc/src/hmac.rs:
```diff
@@ -0,0 +1,256 @@
+        regs.cfg
+            .write(CFG::ENDIAN_SWAP::SET + CFG::SHA_EN::SET + CFG::DIGEST_SWAP::SET);
+
+        let mut key_len = 8;
```
This of course is a subjective comment, but I dislike use of `mut` and often find it confusing. Mostly they can easily be replaced by non mutable variables. Furthermore, redefining a variable should often be preferred instead of making it mutable (e.g. use `let a = a + 1;` instead of `a += 1` where it makes sense).

In this example, I'd recommend something like
```
let key_len =
    if key.len() <= 8 {
        key.len()
    } else {
        8
    };
```
or similar when using `match`. I don't want to be picky here - but reading `mut` in Rust immediately prepares me for a mutable reference or something complex.

gendx replied:
Instead of:
```rust
let key_len =
    if key.len() <= 8 {
        key.len()
    } else {
        8
    };
```
are we allowed to depend on `core` and write the following?
```rust
let key_len = core::cmp::min(key.len(), 8);
```

lschuermann commented on boards/components/src/hmac.rs:
```diff
@@ -0,0 +1,161 @@
+    };};
+}
+
+pub struct HmacsComponent<A: 'static + digest::Digest<'static, T>, T: 'static + digest::DigestType>
```
Why is this called `HmacsComponent` (plural)?

alistair23 replied:
That's what the `LED` one did, I have dropped the s

bradjc commented on capsules/src/hmac.rs:
```diff
@@ -0,0 +1,488 @@
+            // Pass buffer for the key to be in
+            0 => self.appid.map_or(ReturnCode::FAIL, |id| {
+                self.apps
+                    .enter(*id, |app, _| {
```
I think because we have a grant we can remove all of the checks for driver ownership (i.e. `self.appid`) for allow and subscribe. Instead, the buffers and callback can just be unconditionally stored.

bradjc commented on capsules/src/hmac.rs:
```diff
@@ -0,0 +1,488 @@
+//! HMAC (Hash-based Message Authentication Code).
```
This is a tricky `Driver` to write with respect to virtualization. There is a key decision to make:
- Apps have to allow the entire buffer to be HMACd.
or
- Apps can can call "add data" like the HIL allows, and only allow a much smaller buffer than the actual data to be HMACd.

If the latter, we have to allow multiple syscall/callback cycles, yet since the data have to go to the HW, the app effectively "owns" the HMAC HW during those loops. What happens if an app indicates it wants to do a really long HMAC, but then stops providing data? The `Driver` needs some way to reclaim the hardware. I don't think we have another `Driver` currently that presents this issue.

I think for now, using the former option and setting up the driver like our other capsules (ninedof for example) is a reasonable compromise. That would mean restructuring this capsule slightly so that there is an "enqueue_command" function that is triggered on a run command that iteratively executes HMACs it if multiple apps request them.

alistair23 replied:
The idea is that whatever data buffer the app passes to the driver will be HMACd. I don't anticipate a use case where an app only wants to HMAC part of the data buffer. We currently only allow a single data buffer to be used, although that might be changeable in the future.

What I am picturing (untested) is that two apps both create the buffers they want and call the `allow` syscalls to tell the driver. One app will call the `run` syscall first and start the HMAC process. While this is happening if the other app calls `run` they will get an error and have to retry. Once the first app has finished then the second app can call `run` and work successfully.

You think the `run` logic should be changed to allow a queue of run commands so we don't return an error. I'll give that a go.

lschuermann replied:
> You think the `run` logic should be changed to allow a queue of run commands so we don't return an error. I'll give that a go.

I think this is the correct approach. Alternatively, to make the apps aware of the possible delay, we could keep the current approach and introduce a subscription on the event that the driver is available again. That would remove the need for complex queuing and give the apps more control.

alistair23 commented:
I have pushed an update that queues the next run event, this is similar to the buzzer capsule.

I have also changed the `clear_data` API so apps never call `clear_data`. Instead the driver will clear the data when it's finished or if there is an error.

alistair23 commented:
Three lines added :)

gendx commented on chips/lowrisc/src/hmac.rs:
```diff
@@ -0,0 +1,252 @@
+
+                    let idx = i * 4;
+
+                    digest[idx + 0] = d[3];
+                    digest[idx + 1] = d[2];
+                    digest[idx + 2] = d[1];
+                    digest[idx + 3] = d[0];
```
This reverts the byte order of `d`. Couldn't this block be replaced by:

```rust
let d = regs.digest[i].get().swap_bytes().to_ne_bytes();

let idx = i * 4;

digest[idx + 0] = d[0];
digest[idx + 1] = d[1];
digest[idx + 2] = d[2];
digest[idx + 3] = d[3];
```

Or (with whatever endianness is not native):

```rust
let d = regs.digest[i].get().to_(be|le)_bytes();
```

jmichelp replied:
Or maybe all this endianness swapping can be avoided by clearing the `CFG::DIGEST_SWAP` bit instead of setting it?
I'm not familiar with OT and unfortunately the FPGA boards I have are not compatible to try it.

alistair23 replied:
I have removed the endlessness swap

jmichelp commented on kernel/src/hil/digest.rs:
```diff
@@ -0,0 +1,72 @@
+    /// If there is data from the `add_data()` command asyncrously waiting to
+    /// be written it will be written before the operation starts.
+    ///
+    /// If an appropriote `set_mode*()` wasn't called before this function the
```
Typo

jmichelp commented on kernel/src/hil/digest.rs:
```diff
@@ -0,0 +1,72 @@
+    /// There is no guarantee the data has been written until the `add_data_done()`
+    /// callback is fired.
+    /// On error the return value will contain a return code and the original data digest
+    fn add_data(
```
The name of the method as well as the documentation lead me to think that if the data I want to hash doesn't fit in RAM entirely, I can call repeatedly the `add_data` function over several chunks (waiting for the callback between every chunk of course) and then call `run` to get the digest.
If this is the case, then I think I'm missing a `start_new_message` method or something similar. Otherwise the digest will never be correct because the initial value will reloaded at each chunk. If I read the code correctly that's what the current implementation for OT module is doing in this PR:
```rust
regs.cmd.modify(CMD::START::SET);
```

Maybe I'm too used to standard crypto hashing API (`new`, `update`, `finish`).

alistair23 replied:
The documentation and name are correct and you can call `add_data()` multiple times to hash something that doesn't fit into RAM.

Why do you need a `start_new_message()` function?

jmichelp replied:
Ok so if I got it correctly, the `start_new_message()` function I was looking for is actually the`set_mode_hmacsha256()` function. That's how you tell the hardware module that you're going to compute another hash or hmac.
Let's say I need to compute hmac-sha256 with the same key for 2 message, with your HIL, am I supposed to do?

1. `set_mode_hmacsha256(key)`
1. `add_data(msg1_part1)`, `add_data(msg1_part2)`, ...., `add_data(msg1_partN)`
1. `run(mac_msg1)`
1. `set_mode_hmacsha256(key)`
1. `add_data(msg1_part1)`, `add_data(msg1_part2)`, ...., `add_data(msg1_partN)`
1. `run(mac_msg1)`

or is the step number 4 optional?

I'm trying to implement your HIL for the ARM CryptoCell-31x and it requires manually load the initial value for the selected digest algorithm. That's why I'm trying to figure out when to do it. If it is in the `set_mode_hmacsha256()` function, then we have a 1:1 equivalent to the more traditional `new()` (`set_mode_hmacsha256`), `update()` (`add_data()`), `finish()` (`run()`) API I've been mentioning.
If we don't need to call `set_mode_hmacsha256()` between two messages, I need to implement it both in `set_mode_hmacsha256()` and in `run()` functions.

Hope this makes sense :)

alistair23 replied:
Step 4 should be optional in this case (untested though).

I'm happy to update the documentation and implementation to make step 4 compulsory if required.

jmichelp commented on kernel/src/hil/digest.rs:
```diff
@@ -0,0 +1,72 @@
+    /// Returns the number of bytes parsed on success
+    /// There is no guarantee the data has been written until the `add_data_done()`
+    /// callback is fired.
+    /// On error the return value will contain a return code and the original data digest
```
`original data digest` or `original data`?
How could it contain a (valid) digest before we even call `run` and on top of that in case of error?

alistair23 replied:
Whoops, typo here. Fixed.

jmichelp commented on chips/lowrisc/src/hmac.rs:
```diff
@@ -0,0 +1,252 @@
+        regs.cfg
+            .write(CFG::ENDIAN_SWAP::SET + CFG::SHA_EN::SET + CFG::DIGEST_SWAP::SET);
+
+        let key_len = core::cmp::min(key.len(), 8);
```
Something I missed at first. `key` is now a fixed-size array instead of a slice. Comparing the length is useless as `key.len()` will always return 32.

alistair23 commented:
I addressed the few small comments

bradjc commented:
bors r+
